### PR TITLE
robust imports stage 2

### PR DESCRIPTION
**Pull request by bdemann**

### CODE REVIEW

File: src/compiler/typescript_to_rust/azle_generate/src/candid_type/record/mod.rs
Comment (lastmjs) [inline src/compiler/typescript_to_rust/azle_generate/src/candid_type/record/mod.rs:1]: We need to figure out how to optimize by not scanning files entirely if possible if there are no symbols, or not scanning each file for every symbol if some symbols are empty (do not have aliases).

File: src/compiler/typescript_to_rust/azle_generate/src/candid_type/type_ref/mod.rs [hunk 1]
```diff
@@ -69,15 +69,25 @@ impl SourceMapped<'_, TsTypeRef> {
...
-        Ok(match self.get_name()? {
-            "Opt" => CandidType::Opt(self.to_option()?),
-            "Func" => CandidType::Func(self.to_func(None)?),
-            "Record" => CandidType::Record(self.to_record()?),
-            "Tuple" => CandidType::Tuple(self.to_tuple()?),
-            "Vec" => CandidType::Array(self.to_vec()?),
-            "Variant" => CandidType::Variant(self.to_variant()?),
-            _ => CandidType::TypeRef(self.to_type_ref()?),
-        })
+        if let Some(opt) = self.to_option()? {
```
Comment (lastmjs) [inline src/compiler/typescript_to_rust/azle_generate/src/candid_type/type_ref/mod.rs:72]: Let's bring the match back

File: src/compiler/typescript_to_rust/azle_generate/src/candid_type/type_ref/mod.rs [hunk 2]
```diff
@@ -56,7 +56,7 @@ impl SourceMapped<'_, TsTypeRef> {
             .collect_results()?;
 
         Ok(TypeRef {
-            name: if name_string == "Result" {
+            name: if self.symbol_table.result.contains(&name_string.to_string()) {
```
Comment (lastmjs) [inline src/compiler/typescript_to_rust/azle_generate/src/candid_type/type_ref/mod.rs:59]: I think we need to revert this

File: src/compiler/typescript_to_rust/azle_generate/src/candid_type/record/mod.rs [hunk 3]
```diff
@@ -16,25 +19,41 @@ use super::errors::WrongEnclosedType;
...
+        .map(|result| result.flatten())
     }
 }
 
 impl SourceMapped<'_, TsTypeRef> {
-    pub fn to_record(&self) -> Result<Record, Vec<Error>> {
-        match self.get_ts_type()?.as_ts_type_lit() {
-            Some(ts_type_lit) => ts_type_lit,
-            None => return Err(vec![WrongEnclosedType::error_from_ts_type_ref(self).into()]),
+    pub fn to_record(&self) -> Result<Option<Record>, Vec<Error>> {
```
Comment (lastmjs) [inline src/compiler/typescript_to_rust/azle_generate/src/candid_type/record/mod.rs:39]: Revert

File: src/compiler/typescript_to_rust/azle_generate/src/candid_type/variant/mod.rs [hunk 4]
```diff
@@ -16,34 +19,52 @@ use super::errors::WrongEnclosedType;
...
+        self.process_ts_type_ref(&self.symbol_table.variant, |type_ref| {
             // TODO this should be undone once we put all user-defined types in their own module
             let name_string = self.id.get_name().to_string();
-            let name = Some(if name_string == "Result" {
-                "_AzleResult".to_string()
-            } else {
-                name_string
-            });
+            let name = Some(
+                if self.symbol_table.result.contains(&name_string.to_string()) {
```
Comment (lastmjs) [inline src/compiler/typescript_to_rust/azle_generate/src/candid_type/variant/mod.rs:26]: All of these "Result" things should be left alone

Comment (lastmjs) [inline src/compiler/typescript_to_rust/azle_generate/src/candid_type/variant/mod.rs:26]: Result should be removed from the symbol table (I think, Ben seems sure).

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: src/compiler/typescript_to_rust/azle_generate/src/candid_type/record/mod.rs (hunk 3 above)
Comment (bdemann) [reply src/compiler/typescript_to_rust/azle_generate/src/candid_type/record/mod.rs:39]: We decided to leave Result<Option> with the thought that we might change to Option<Result>

File: src/compiler/typescript_to_rust/azle_generate/src/candid_type/type_ref/mod.rs (hunk 1 above)
Comment (bdemann) [reply src/compiler/typescript_to_rust/azle_generate/src/candid_type/type_ref/mod.rs:73]: We we are moving to Result<Option> then the if let some is going to be the most concise way to handle this.

File: src/compiler/typescript_to_rust/azle_generate/src/body/stable_b_tree_map/new_expr/errors/mod.rs [hunk 5]
```diff
@@ -21,7 +21,7 @@ pub fn build_type_arg_error_message(
     location: &Location,
 ) -> CompilerOutput {
     let range = (
-        // UNWRAP HERE
+        // UNWRAP HERE // TODO this will be made more complicated by robust imports
```
Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/body/stable_b_tree_map/new_expr/errors/mod.rs:24]: comment this out

File: src/compiler/typescript_to_rust/azle_generate/src/candid_type/errors/wrong_enclosed_type.rs [hunk 6]
```diff
@@ -3,35 +3,49 @@ use swc_ecma_ast::TsTypeRef;
...
         };
-        let (modified_source, modified_range) = match sm_ts_type_ref.wrong_enclosed_type_error() {
-            Ok(value) => value,
-            Err(err) => return err,
+        let (modified_source, modified_range) = match name {
+            EnclosingType::Variant => match sm_ts_type_ref.variant_wrong_enclosed_type_error() {
+                Ok(value) => value,
+                Err(err) => return err,
+            },
+            EnclosingType::Func => sm_ts_type_ref.get_func_suggestion_modifications_and_stuff(),
```
Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/candid_type/errors/wrong_enclosed_type.rs:46]: rename this stuff

File: src/compiler/typescript_to_rust/azle_generate/src/candid_type/mod.rs [hunk 7]
```diff
@@ -42,13 +42,25 @@ impl TsAst {
         })
     }
 
-    pub fn extract_candid_types<F, T>(&self, extractor: F) -> Result<Vec<T>, Vec<Error>>
+    pub fn extract_candid_types<G, F, T, Symbol>(
+        &self,
+        extractor: F,
+        ask_dan_and_jordan_for_better_namer: G,
```
Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/candid_type/mod.rs:48]: rename to "get_symbol_alias"

File: src/compiler/typescript_to_rust/azle_generate/src/candid_type/mod.rs [hunk 8]
```diff
@@ -23,12 +23,12 @@ pub mod vec;
 impl TsAst {
     pub fn build_candid_types(&self) -> Result<CandidTypes, Vec<Error>> {
         let (type_aliases, funcs, records, services, tuples, variants) = (
-            self.extract_candid_types(|x| x.to_type_alias()),
-            self.extract_candid_types(|x| x.to_func()),
-            self.extract_candid_types(|x| x.to_record()),
+            self.extract_candid_types(|x| x.to_type_alias(), |p| &p.symbol_table.alias),
```
Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/candid_type/mod.rs:26]: rename x and p

Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/candid_type/mod.rs:26]: Get rid of this optimization 
Make a ticket with this option in it
add //TODO performance optimization opportunity

File: src/compiler/typescript_to_rust/azle_generate/src/candid_type/record/mod.rs [hunk 9]
```diff
@@ -16,25 +19,41 @@ use super::errors::WrongEnclosedType;
 
 impl SourceMapped<'_, TsTypeAliasDecl> {
     pub fn to_record(&self) -> Result<Option<Record>, Vec<Error>> {
-        self.process_ts_type_ref("Record", |type_ref| {
-            let (type_params, members) =
+        self.process_ts_type_ref(&self.symbol_table.record, |type_ref| {
+            let (type_params, record_right_half) =
```
Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/candid_type/record/mod.rs:23]: rename to "right_hand_side" or "members"
actually call it "type_ref"
type Coolboy = Record<{}>

File: src/compiler/typescript_to_rust/azle_generate/src/candid_type/mod.rs (hunk 7 above)
Comment (bdemann) [reply src/compiler/typescript_to_rust/azle_generate/src/candid_type/mod.rs:48]: We are not doing this optimization so this rename fizzles

File: src/compiler/typescript_to_rust/azle_generate/src/candid_type/mod.rs (hunk 8 above)
Comment (bdemann) [reply src/compiler/typescript_to_rust/azle_generate/src/candid_type/mod.rs:26]: we aren't doing this optimization so this rename fizzles

**Review by lastmjs — APPROVED** [review]

**Pull request merged**